### PR TITLE
ARROW-10675 [C++][Python] Support AWS S3 Web identity credentials

### PR DESCRIPTION
**Pull request by sahil1105**
Added basic support for explicitly specifying use of web identity credentials (it's already part of the default credentials provider chain) in the C++ API. Will add Python API in a separate PR.
Also refactored some existing code to make serialization and deserialization in Python more structured by adding an enum class `S3CredentialsKind` which uniquely identifies the credential type being used.
Also adds some missing error-checking and tests in the Python API.

### CODE REVIEW

**Comment by sahil1105** [comment]
@pitrou Could you review when you get a chance? Thanks.

File: cpp/src/arrow/filesystem/s3fs.h
Comment (pitrou) [inline cpp/src/arrow/filesystem/s3fs.h:101]: Aren't `anonymous`, `creds_provided` and `use_web_identity` all exclusive? If so, how about replacing them with an enum, e.g.:
```c++
struct S3Options {
  enum CredentialsKind {
    /// Anonymous access (no credentials used)
    Anonymous,
    /// Use default AWS credentials, configured through environment variables
    Default,
    /// Use explicitly-provided username and password
    Explicit,
    /// Use web identity token to assume role, configured through environment variables
    WebIdentity
  };

  CredentialsKind credentials_kind = Default;
```

Comment (sahil1105) [reply cpp/src/arrow/filesystem/s3fs.h:101]: Yes, that'a good idea. Thanks.

Comment (sahil1105) [reply cpp/src/arrow/filesystem/s3fs.h]: @pitrou Done. Apologies for the delay.

File: python/pyarrow/_s3fs.pyx
Comment (pitrou) [inline python/pyarrow/_s3fs.pyx:83]: Hmm... I think it would be better to avoid adding boolean arguments for each different kind of authentication.
I'm not sure what a good API would be, but perhaps we could add a generic `auth` parameter that would accept the following values:
* `auth='default'`
* `auth='anonymous'`
* `auth={'access_key': '...', 'secret_key': '...', 'session_token': '...'}`
* `auth={'role_arn': '...', 'session_name': '...', 'external_id': '...', 'load_frequency': 123}`
* `auth='web_identity'`

@jorisvandenbossche What do you think?

Comment (jorisvandenbossche) [reply python/pyarrow/_s3fs.pyx:83]: Without knowing too much about S3 / AWS auth, your proposal feels quite natural for expressing the different options, and could be a nice alternative to a bunch of partially exclusive keyword arguments each for a different way of authorization. We would (initially) need to keep the keywords for backward compatibility.

Comment (nealrichardson) [reply python/pyarrow/_s3fs.pyx:83]: Whatever you decide, could you please document and write up a JIRA for doing the same in the R package (assuming you're not going to go all the way and update the R bindings in this PR, though you're more than welcome to :)

Comment (sahil1105) [reply python/pyarrow/_s3fs.pyx:83]: @pitrou That makes sense to me as well.
I think it might be better to do that in a separate PR though (to keep the scope of this one limited).
Let me know.

Comment (pitrou) [reply python/pyarrow/_s3fs.pyx:83]: I would be ok with doing the Python changes in a separate PR, but then you should remove them from this PR ;-)

Comment (jorisvandenbossche) [reply python/pyarrow/_s3fs.pyx:83]: @sahil1105 you could also do a pre-cursor PR to refactor the existing options like discussed here

Comment (sahil1105) [reply python/pyarrow/_s3fs.pyx:83]: Wouldn't we still need the python changes in this PR even if we move to using a generic `auth` parameter (since we need to keep the keywords around for now for backward compatibility)? Or am I missing something?

Comment (jorisvandenbossche) [reply python/pyarrow/_s3fs.pyx:83]: The Python changes in this PR are not needed for backwards compatibility? (since those are new options) 
So assuming we refactor the python options to use a generic `auth` parameter, then this PR with new functionality should directly use that new parameter, and not introduce a new `use_web_identity` keyword that would then later need to be deprecated. 
So for this PR, I think either remove any Python bindings in this PR (and keep this PR to only add the C++ functionality, and later do a follow-up PR to add the python bindings) or either first do a precursor PR to refactor the options to use `auth` (and then afterwards this PR can be updated to add the web identity functionality into an `auth` keyword).

Comment (sahil1105) [reply python/pyarrow/_s3fs.pyx:83]: I agree about removing the `use_web_identity` keyword arg. I meant all the other Python changes. Those are actually just simple error-checking changes that we should keep in my opinion.
Related Q: Should I remove `S3CredentialsKind` and related changes? I think it's good to have them, but I'll defer to you regarding whether to keep them.

Comment (pitrou) [reply python/pyarrow/_s3fs.pyx:83]: I agree that it's ok to keep the additional error-checking changes, as well as the `S3CredentialsKind` declaration.

Comment (sahil1105) [reply python/pyarrow/_s3fs.pyx]: Done. Sorry for the delay.

**Comment by sahil1105** [comment]
@pitrou @jorisvandenbossche I think this is ready. Let me know if any other changes need to be made. Once this is merged I can start on a PR to add the auth keyword argument we discussed (and deprecate the others).

**Review by pitrou — APPROVED** [review]
Sorry for the delay @sahil1105 . This looks good to me, thank you!

**Comment by rngtng** [comment]
Thanks for this feature! I require this now for ruby, but can't get it working. Any hints how to activate it?

 So far I follow  tis https://github.com/apache/arrow/tree/master/ruby#from-s3  and ensure to have this ENV set:
`AWS_DEFAULT_REGION, AWS_ROLE_ARN, AWS_WEB_IDENTITY_TOKEN_FILE and AWS_ROLE_SESSION_NAME`. But I guess I explicit have to set `credentials_kind`? how can I do this in ruby? 

Thanks, Tobi

**Comment by pitrou** [comment]
@kou  Would know more about the Ruby bindings.

**Comment by kou** [comment]
@rngtng It seems that we need to implement bindings of this explicitly for Ruby. We can't use this with environment variables.
Could you open an issue on JIRA for implementing bindings of this for Ruby?
https://issues.apache.org/jira/browse/ARROW

**Pull request closed without merging**
